File: leetcode3/최원준/3684. Maximize Sum of At Most K Distinct Elements.py
#

'''
1. 아이디어 :
정렬 후, set을 사용하여 방문한 숫자들을 체크

2. 시간복잡도 :
    O(nlogn)

3. 자료구조/알고리즘 :
해시셋, 정렬

'''
class Solution:
    def maxKDistinct(self, nums: List[int], k: int) -> List[int]:
        visited = set()
        nums = sorted(nums, reverse = True)
        
        ans = []
        counts = 0

        for num in nums:
            if num in visited:
                continue
            
            ans.append(num)
            visited.add(num)
            counts += 1
            if counts == k:
                return ans
        return ans

File: leetcode3/이진희/3684_Maximize Sum of At Most K Distinct Elements.java
/*

1. 아이디어 : 정렬후, Set을 사용해 파악

2. 시간복잡도 : O(NlogN) + O(N) = O(NlogN)

3. 자료구조/알고리즘 : 정렬, Set

 */

class Solution {
    public int[] maxKDistinct(int[] nums, int k) {
        //내림차 순 반환
        //중복 선택 안됨

        int[] ans = new int[k];
        int idx = 0;
        Set<Integer>set = new HashSet<>();

        Arrays.sort(nums);
        for(int i=nums.length-1; i>=0; i--) {
            if(set.size()==k) break;
            if(set.contains(nums[i])) continue;
            set.add(nums[i]);
            ans[idx++] = nums[i];
        }

        return Arrays.copyOfRange(ans,0,idx);
    }
}